### PR TITLE
Adding milestone, lock, and block-stage steps to the demo

### PR DESCRIPTION
**Pull request by amuniz**
Downstream of 

- https://github.com/jenkinsci/pipeline-milestone-step-plugin/pull/1
- https://github.com/jenkinsci/lockable-resources-plugin/pull/25 
- https://github.com/jenkinsci/pipeline-stage-step-plugin/pull/4

@reviewbybees

### CODE REVIEW

File: demo/Dockerfile [hunk 1]
```diff
@@ -29,7 +29,7 @@ USER jenkins
 RUN echo '<settings><mirrors><mirror><id>central</id><url>http://repo.jenkins-ci.org/simple/repo1-cache/</url><mirrorOf>central</mirrorOf></mirror></mirrors><localRepository>/usr/share/jenkins/ref/.m2/repository</localRepository></settings>' > settings.xml
 RUN /usr/local/maven/bin/mvn -s settings.xml -f repo-wc install && /usr/local/maven/bin/mvn -s settings.xml -f repo-wc/sometests -Dmaven.test.failure.ignore clean install
 
-COPY plugins.txt .
+COPY plugins.txt ./
```
Comment (amuniz) [inline demo/Dockerfile]: Not sure how it worked before. Without the slash I get [this](http://stackoverflow.com/questions/26405702/dockerfile-copy-lchown-not-a-directory).

File: demo/Makefile [hunk 2]
```diff
@@ -11,7 +11,7 @@ run: build
 build-snapshot:
 	docker build -t $(IMAGE):RELEASE .
 	mkdir -p snapshot-plugins
-	for p in $$(cat plugins.txt|perl -pe s/:.+//g; cat snapshot-only-plugins.txt); do echo looking for snapshot builds of $$p; for g in org/jenkins-ci/plugins org/jenkins-ci/plugins/workflow org/jenkins-ci/plugins/pipeline-stage-view; do if [ -f ~/.m2/repository/$$g/$$p/maven-metadata-local.xml ]; then cp -v $$(ls -1 ~/.m2/repository/$$g/$$p/*-SNAPSHOT/*.hpi | tail -1) snapshot-plugins/$$p.jpi; fi; done; done
+	for p in $$(cat snapshot-only-plugins.txt); do echo looking for snapshot builds of $$p; for g in org/jenkins-ci/plugins org/jenkins-ci/plugins/workflow org/jenkins-ci/plugins/pipeline-stage-view; do if [ -f ~/.m2/repository/$$g/$$p/maven-metadata-local.xml ]; then cp -v $$(ls -1 ~/.m2/repository/$$g/$$p/*-SNAPSHOT/*.hpi | tail -1) snapshot-plugins/$$p.jpi; fi; done; done
```
Comment (amuniz) [inline demo/Makefile]: This is not really related, but it was really annoying to have every snapshot in my local Maven repo deployed in the snapshot docker image (`make build-snapshot`). I would expect to have snapshots deployed *only* for the plugins I explicitly add in `snapshot-only-plugins.txt`.

File: demo/snapshot-only-plugins.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+pipeline-milestone-step
```
Comment (amuniz) [inline demo/snapshot-only-plugins.txt]: To be removed once released.

File: demo/Makefile (hunk 2 above)
Comment (jglick) [inline demo/Makefile]: :bug: this behavior was intentional. It is the only way to test the behavior of unreleased changes in the image.

`snapshot-only-plugins.txt` is for plugins which have _never_ been released.

File: demo/plugins.txt [hunk 4]
```diff
@@ -9,7 +9,10 @@ git-server:1.6
 handlebars:1.1.1
 icon-shim:2.0.3
 jquery-detached:1.2.1
+lockable-resources:1.8
+mailer:1.5
```
Comment (jglick) [inline demo/plugins.txt]: Why?

File: demo/plugins.txt [hunk 5]
```diff
@@ -9,7 +9,10 @@ git-server:1.6
...
 mapdb-api:1.0.6.0
+matrix-project:1.4
```
Comment (jglick) [inline demo/plugins.txt]: Why?

File: demo/snapshot-only-plugins.txt (hunk 3 above)
Comment (jglick) [inline demo/snapshot-only-plugins.txt]: Rather, moved into `plugins.txt`.

File: demo/repo/Jenkinsfile [hunk 6]
```diff
@@ -17,26 +17,31 @@ parallel(longerTests: {
     runTests(servers, 20)
 })
 
-stage name: 'Staging', concurrency: 1
-node {
-    servers.deploy 'staging'
-}
+milestone ordinal: 1
```
Comment (jglick) [inline demo/repo/Jenkinsfile]: Should suffice to write

```groovy
milestone 1
```

File: demo/repo/Jenkinsfile [hunk 7]
```diff
@@ -17,26 +17,31 @@ parallel(longerTests: {
...
+lock (resource: 'staging-server', inversePrecedence: true) {
```
Comment (jglick) [inline demo/repo/Jenkinsfile]: For consistency with other step usages (and _Snippet Generator_ output), please elide the space before the opening parenthesis. See for example `dir` above.

File: demo/repo/Jenkinsfile [hunk 8]
```diff
@@ -17,26 +17,31 @@ parallel(longerTests: {
...
+    stage name: 'Staging'
```
Comment (jglick) [inline demo/repo/Jenkinsfile]: Probably you would want this _outside_ the `lock`, so that in the stage view you would see that we are _waiting_ to stage, which I think is how the previous script worked. (The stage view will show delays explicitly if you add `PauseAction` from the `lock` step.)

Comment (jglick) [inline demo/repo/Jenkinsfile]: cf. [JENKINS-27129](https://issues.jenkins-ci.org/browse/JENKINS-27129)

File: demo/repo/Jenkinsfile [hunk 9]
```diff
@@ -17,26 +17,31 @@ parallel(longerTests: {
...
+    }
 }
 
-stage name: 'Production', concurrency: 1
-node {
-    sh "wget -O - -S ${jettyUrl}staging/"
-    echo 'Production server looks to be alive'
-    servers.deploy 'production'
-    echo "Deployed to ${jettyUrl}production/"
+milestone()
```
Comment (jglick) [inline demo/repo/Jenkinsfile]: Why not

```groovy
milestone 2
```

since we specified `1` before?

File: demo/repo/Jenkinsfile [hunk 10]
```diff
@@ -17,26 +17,31 @@ parallel(longerTests: {
...
+lock ('production-server') {
+    stage name: 'Production'
```
Comment (jglick) [inline demo/repo/Jenkinsfile]: ditto

File: demo/Makefile (hunk 2 above)
Comment (amuniz) [inline demo/Makefile]: I expected this bug TBH :)

So, as any developer, I have plenty of plugin snapshots in my local Maven repository which I just want to ignore. Is it supposed that I have to clean my local `.m2/repository` if I want to use only released versions?

> It is the only way to test the behavior of unreleased changes in the image

If you want to test plugin `my-plugin2.3-SNAPSHOT`: run `mvn clean install`, add `my-plugin` to `snapshot-only-plugins.txt` and rebuild the image. I think it worked for me. Am I missing something?

File: demo/plugins.txt (hunk 4 above)
Comment (amuniz) [inline demo/plugins.txt]: Transitive dependencies of `lockable-resources`. All dependencies must be explicitly set as there is no dependencies resolution here, no?

File: demo/plugins.txt (hunk 5 above)
Comment (amuniz) [inline demo/plugins.txt]: Transitive dependency of `lockable-resources`

File: demo/repo/Jenkinsfile (hunk 6 above)
Comment (amuniz) [inline demo/repo/Jenkinsfile]: Yeah, will fix.

File: demo/repo/Jenkinsfile (hunk 7 above)
Comment (amuniz) [inline demo/repo/Jenkinsfile]: Ok.

File: demo/repo/Jenkinsfile (hunk 8 above)
Comment (amuniz) [inline demo/repo/Jenkinsfile]: Yeah, I'm changing the previous behavior. IMO the build is not really *in* _Staging_ until it acquires the lock.

> (The stage view will show delays explicitly if you add PauseAction from the lock step.

Thanks for the interesting tip! I'll add this to `lock` (but not now).

File: demo/repo/Jenkinsfile (hunk 9 above)
Comment (amuniz) [inline demo/repo/Jenkinsfile]: Yes. I just wanted to show that it's possible to use mixed _defined_ and _non-defined_ ordinals in the same script (as it was kind of controversial in the `milestone` PR discussion).

File: demo/Makefile (hunk 2 above)
Comment (jglick) [inline demo/Makefile]: The intention, and the original behavior of `workflow-plugin/demo/Dockerfile#run-snapshot`, was that you should not need to modify demo source files in order to pick up snapshots. If you are running a snapshot demo anyway, you are generally interested in knowing about any problems introduced by unreleased changes.

Now we could amend that to require you to list the plugins to be searched in a variable passed to `make`, e.g.

    make -C demo run-snapshot SNAPSHOT_PLUGINS='cloudbees-folder branch-api workflow-multibranch'

where `snapshot-only-plugins.txt` could be deleted in favor of temporary edits to the default value of this variable in the `Makefile`.

At any rate, this is better worked out in a separate PR.

File: demo/plugins.txt (hunk 4 above)
Comment (jglick) [inline demo/plugins.txt]: > there is no dependencies resolution here

Correct.

File: demo/repo/Jenkinsfile (hunk 8 above)
Comment (jglick) [inline demo/repo/Jenkinsfile]: > the build is not really in _Staging_ until it acquires the lock

Well the build is certainly not in _QA_. To my mind the original behavior, of starting to show the _Staging_ table cell but pausing there, is more informative, though `PauseAction` would make it clearer.

What we get when we add `label` is another question—we have not yet decided how to render sections of the build that are not inside a labeled block, if there are any. I still feel that we would want `label` outside `lock`, with a `PauseAction` coming from lock, to make it clear that we are _trying_ to stage and just being held up.

File: demo/repo/Jenkinsfile (hunk 9 above)
Comment (jglick) [inline demo/repo/Jenkinsfile]: Understood, but that is better left to a functional test; this demo is something we expect users to study and copy from, so it ought to show our recommended style.

File: demo/repo/Jenkinsfile (hunk 6 above)
Comment (amuniz) [inline demo/repo/Jenkinsfile]: Ups, sorry. No, it does not. As both parameters (`label` and `ordinal`) are optional, then the name has to be explicitly set.

Comment (jglick) [inline demo/repo/Jenkinsfile]: Hence my comment in the upstream PR.

File: demo/Dockerfile (hunk 1 above)
Comment (svanoort) [inline demo/Dockerfile]: What version of Docker are you using?  I've seen a couple quirks around this where behaviors changed between docker versions.

File: demo/Makefile (hunk 2 above)
Comment (svanoort) [inline demo/Makefile]: 🐜 as well, I think worth a separate PR too.

File: demo/repo/Jenkinsfile (hunk 9 above)
Comment (svanoort) [inline demo/repo/Jenkinsfile]: Maybe as a comment inline here?  I like the notion that we show best practices, but it's also helpful to know this syntax is valid too.

File: demo/repo/Jenkinsfile [hunk 11]
```diff
@@ -17,26 +17,32 @@ parallel(longerTests: {
...
 }
 
-stage name: 'Production', concurrency: 1
-node {
-    sh "wget -O - -S ${jettyUrl}staging/"
-    echo 'Production server looks to be alive'
-    servers.deploy 'production'
-    echo "Deployed to ${jettyUrl}production/"
+milestone()
+lock('production-server') {
```
Comment (svanoort) [inline demo/repo/Jenkinsfile]: 🐜 Similar to previous comment: one consistent style is helpful, alternately an explaining comment like "this is the same as specifying resource: 'production-server' "

File: demo/repo/Jenkinsfile [hunk 12]
```diff
@@ -17,26 +17,32 @@ parallel(longerTests: {
     runTests(servers, 20)
 })
 
-stage name: 'Staging', concurrency: 1
-node {
-    servers.deploy 'staging'
+milestone ordinal: 1
```
Comment (jglick) [inline demo/repo/Jenkinsfile]: :ant: as noted upstream, `ordinal: ` should be optional.

**Comment by amuniz** [comment]
@jglick updated.

**Comment by jglick** [comment]
Needs to be updated to include https://github.com/jenkinsci/pipeline-stage-step-plugin/pull/4.

File: demo/repo/Jenkinsfile (hunk 9 above)
Comment (jglick) [inline demo/repo/Jenkinsfile]: :ant: the demo should show explicit ordinals:

```groovy
milestone 1
// …
milestone 2
```

**Comment by amuniz** [comment]
@jglick I found some cases where the blocked `stage` can not cover what non-blocked `stage` covered. For example:

```
stage 'QA'
echo 'Hi!'
lock(resource: 'staging-server') {
    stage name: 'Staging'
    echo 'In staging'
}
```

The action to acquire _staging-server_ is in _QA_ stage, but I see no way to do that with blocked `stage`s. Or maybe `stage name: 'Staging'` should be conceptually before `lock`?

**Comment by jglick** [comment]
@amuniz what is wrong with (roughly)

```groovy
stage('Staging') {
  lock('staging-server') {
    node {
      servers.deploy 'staging'
    }
    input "Does ${jettyUrl}staging/ look good?"
  }
}
checkpoint 'Before production'
```

?

**Comment by amuniz** [comment]
Nothing wrong, but not equivalent. The build is in `Staging` stage while waiting to acquire the lock, but in the original script it is in `QA` stage.

**Comment by jglick** [comment]
Well then we have fixed a buglet: logically it makes no sense for the build to be considered in QA when we have finished QA and are now just waiting to stage it.

**Comment by jglick** [comment]
I mean, depending on how the stage view visualizes things, this would also be valid:

```groovy
lock('staging-server') {
  stage('Staging') {
    node {
      servers.deploy 'staging'
    }
    input "Does ${jettyUrl}staging/ look good?"
  }
}
checkpoint 'Before production'
```

which would mean that if we have to wait for the server, there will be a `PauseAction` (`lock` needs to add this BTW if you have not done so already!) which will be _between_ stages rather than _inside_ some stage. I have no strong opinion about which is “better” from a UI rendering standpoint. Keeping `stage` outside `lock` would minimize the need for refinements to the existing visualization scheme.

File: demo/repo/Jenkinsfile [hunk 13]
```diff
@@ -2,41 +2,49 @@ jettyUrl = 'http://localhost:8081/'
...
 
-stage name: 'Production', concurrency: 1
-node {
-    sh "wget -O - -S ${jettyUrl}staging/"
-    echo 'Production server looks to be alive'
-    servers.deploy 'production'
-    echo "Deployed to ${jettyUrl}production/"
+milestone 2
+stage ('Production') {
+    lock('production-server') {
```
Comment (jglick) [inline demo/repo/Jenkinsfile]: Is there a reason `inversePrecedence` is not being specified here too?

**Comment by jglick** [comment]
So. One obvious problem is that @svanoort has not yet patched `pipeline-stage-view` to display blocky stages, so you are back to textual console log visualization (or _Pipeline Steps_).

More to the point, the demo as written does not solve the problem originally stated in JENKINS-27039. If you decline to approve the first build of `cd/master`

```
…
[Pipeline] }
[Pipeline] // stage
[Pipeline] milestone
Trying to pass milestone 1
[Pipeline] stage
[Pipeline] { (Staging)
[Pipeline] lock
Resource [staging-server] did not exist. Created.
Trying to acquire lock on [staging-server]
Lock acquired on [staging-server]
[Pipeline] {
[Pipeline] node
Running on mock-slave-4 in /var/jenkins_home/mock-slaves/mock-slave-4/workspace/cd/master
[Pipeline] {
[Pipeline] unstash
[Pipeline] sh
[master] Running shell script
+ cp x.war /tmp/webapps/staging.war
[Pipeline] }
[Pipeline] // node
[Pipeline] input
Does http://localhost:8081/staging/ look good?
Proceed or Abort
```

and then trigger a second build, it hangs:

```
…
[Pipeline] }
[Pipeline] // stage
[Pipeline] milestone
Trying to pass milestone 1
[Pipeline] stage
[Pipeline] { (Staging)
[Pipeline] lock
Trying to acquire lock on [staging-server]
[staging-server] is locked, waiting...
```

I thought the idea was that once the second build passed milestone 1, if the first build was still between 1 and 2 it would be canceled. But this does not seem to happen.

**Comment by amuniz** [comment]
> So. One obvious problem is that @svanoort has not yet patched pipeline-stage-view to display blocky stages, so you are back to textual console log visualization (or Pipeline Steps).

Yeah, I see this in the logs, maybe unrelated:
```
May 20, 2016 10:08:30 AM org.kohsuke.stapler.framework.adjunct.AdjunctsInPage findNeeded
WARNING: No such adjunct found: org.jenkinsci.pipeline.stageview_adjunct
org.kohsuke.stapler.framework.adjunct.NoSuchAdjunctException: Neither org.jenkinsci.pipeline.stageview_adjunct.css, .js, .html, nor .jelly were found
```

> I thought the idea was that once the second build passed milestone 1, if the first build was still between 1 and 2 it would be canceled. But this does not seem to happen.

Yes, that's what `milestone` is supposed to do, but `input` inside `lock` is preventing the second build to go ahead. So in this usage example, `input` should go outside the `lock`ed block.

Fixed in last commit.

**Comment by jglick** [comment]
> maybe unrelated

I think unrelated.

> `input` inside `lock` is preventing the second build to go ahead

Build 1 is past milestone 1 and holding the lock and paused in input. Build 2 is also past milestone 1. OK, rereading the description of the `milestone` step (BTW the third point is not grammatically correct English and I am not sure what is means), I see that the fourth point says that build 1 will only be aborted once build 2 hits the _second_ milestone, whereas I thought it would be aborted once build 2 hits the _first_ milestone.

But now there may be a problem in that the lock is not held while we prompt for input, so someone could come to this prompt while the server is being updated to some newer build, which would be misleading. What we would really want (I think) is that only one build at a time can be in staging, and if build 1 is waiting in this stage and build 2 arrives, build 1 is aborted then.

At any rate, I should play with the behavior as of the revised demo.

**Comment by amuniz** [comment]
>  I see that the fourth point says that build 1 will only be aborted once build 2 hits the second milestone, whereas I thought it would be aborted once build 2 hits the first milestone.

Right, that way the _never deploy syndrome_ (newer builds always canceling the previous one in a loaded pipeline) is avoided. And this feature is what fixes [JENKINS-27039](https://issues.jenkins-ci.org/browse/JENKINS-27039).

> But now there may be a problem in that the lock is not held while we prompt for input, so someone could come to this prompt while the server is being updated to some newer build, which would be misleading. 

Yes, I think I misread your previous comment about the hanging build. The [previous version of the Jenkinsfile](https://github.com/amuniz/workflow-aggregator-plugin/blob/9fb810e44e960a0335d1c114f4914db0393fe939/demo/repo/Jenkinsfile) was ok and works as expected (locking the `stage-server` and unlocking on build abort/proceed). I'm not able to reproduce the hanging on lock you pointed before.

I'm going back to that version.

**Comment by jglick** [comment]
> I'm not able to reproduce the hanging on lock you pointed before.

No? I thought this was the designed behavior when the `input` was inside `lock`. Build 1 would wait for something to happen, and build 2 would just sit waiting for the lock. Perhaps that is not wrong; it just means that if you do nothing about build 1 but build 3 comes along, 2 will be canceled, and 3 will be waiting for 1. Eventually you either approve or cancel 1; either way, 3 begins staging.

**Comment by amuniz** [comment]
> No?

No. Maybe we are talking about different things. Correct me if I'm wrong, the use case you exposed in [this comment](https://github.com/jenkinsci/workflow-aggregator-plugin/pull/3#issuecomment-220437332) was:

 1. Run build 1
 2. Decline it while waiting on `input`
 3. Run build 2
 4. It hangs waiting for lock `staging-server`

That's not reproducible for me, and if it were, it would be a bug.

> Build 1 would wait for something to happen, and build 2 would just sit waiting for the lock

Yes, this is right.

> it just means that if you do nothing about build 1 but build 3 comes along, 2 will be canceled, and 3 will be waiting for 1

Not exactly. In that situation build 2 and 3 will keep queued until build 1 releases the lock (allowed to proceed or not), at that point the newer build waiting for lock will be allowed to go in (as `inversePrecedence: true` is configured in `lock`), so build 3 acquires the lock. Build 2 will sit waiting for lock. **Only** when build 3 reaches the next milestone then build 2 is cancelled (and it's fine, as build 3 could fail before reaching the next milestone so build 2 would become the newest build which is potentially _fine_ to go to production).

**Comment by jglick** [comment]
No, my use case was

1. run build 1
1. wait for build 1 to ask for input (do not accept nor decline)
1. run build 2
1. build 2 will wait for the lock

Which I suppose is OK.

> Only when build 3 reaches the next milestone then build 2 is cancelled

Note that in the meantime build 2 would probably have acquired the staging lock and started staging, so it will get canceled in the middle of that somewhere.

**Comment by amuniz** [comment]
Then yes, that's ok, build 2 has to wait for the lock.

> Note that in the meantime build 2 would probably have acquired the staging lock

Right. I think it does matter (in this example), otherwise the script needs to be modified as:

```groovy
milestone 1
stage('Staging') {
    lock(resource: 'staging-server', inversePrecedence: true) {
        node {
            servers.deploy 'staging'
        }
        input message: "Does ${jettyUrl}staging/ look good?"
        milestone 2
    }
    ...
}
```

So any older build waiting in `lock` will be cancelled before the lock is released.
I can change it if you feel strong about it.

**Comment by jglick** [comment]
> So any older build waiting in `lock` will be cancelled before the lock is released.

Yeah, worth exploring.

Did you check behavior with checkpoints by the way?

**Comment by amuniz** [comment]
> Did you check behavior with checkpoints by the way?

Yes, basic tests though, it seems to work. Running a checkpoint inside a lock is not reliable (as it was before with `stage concurrency: 1` I believe).
